pkg/token/serviceprincipaltoken.go

```diff
@@ -1,26 +1,42 @@
 package token
 
 import (
+	"crypto/rsa"
+	"crypto/x509"
+	"encoding/pem"
 	"errors"
 	"fmt"
+	"io/ioutil"
 
 	"github.com/Azure/go-autorest/autorest/adal"
+	"golang.org/x/crypto/pkcs12"
+)
+
+//pem block types
+const (
+	certificate = "CERTIFICATE"
+	privateKey  = "PRIVATE KEY"
+)
+
+const (
+	defaultEnvironment = "AzurePublicCloud"
 )
 
 type servicePrincipalToken struct {
 	clientID     string
 	clientSecret string
+	clientCert   string
 	resourceID   string
 	tenantID     string
 	oAuthConfig  adal.OAuthConfig
 }
 
-func newServicePrincipalToken(oAuthConfig adal.OAuthConfig, clientID, clientSecret, resourceID, tenantID string) (TokenProvider, error) {
+func newServicePrincipalToken(oAuthConfig adal.OAuthConfig, clientID, clientSecret, clientCert, resourceID, tenantID string) (TokenProvider, error) {
 	if clientID == "" {
 		return nil, errors.New("clientID cannot be empty")
 	}
-	if clientSecret == "" {
-		return nil, errors.New("clientSecret cannot be empty")
+	if clientSecret == "" && clientCert == "" {
+		return nil, errors.New("Both clientSecret and clientcert cannot be empty")
 	}
```

Comment:
can we return error when both secret and cert are set?

Reply:
done